hashicorp/consul-k8s
# Support upgrades for connect refactor

Before the connect refactor, service registration in Consul was managed by the lifecycle sidecar, which would re-register the service with Consul every 10s. Now, service registration is managed by Endpoints controller.

In order to support upgrades to the refactored Endpoints controller, we need Endpoints controller to NOT register or deregister any services managed by lifecycle sidecar. To do this, the annotation `consul.hashicorp.com/connect-inject-managed-by` is added to pods by the mutating webhook, so endpoints controller will ignore older services managed by lifecycle sidecar (legacy services) for service registration/deregistration. 

To make sure endpoints controller only deregisters services managed by endpoints controller, a meta key indicating the service is managed by endpoints controller is added to the Consul service registration. 

To support health checks for legacy services, the Endpoints controller will always update the healthcheck for any pod, whether it's managed by Endpoints controller or not.

How I've tested this PR:
* Helm install v0.31.1 with the following values
  * <details>
  </details>
* k apply -f static-client.yaml
  * <details>
  </details>
* k apply -f static-server.yaml
  * <details>
  </details>
* Helm install latest with the following values
  * <details>
  </details>
* Check that the services/checks still exist: 
```
k exec pod/nitya-consul-9hhn7 -- curl -s localhost:8500/v1/agent/services?ns=foo | jq .
k exec pod/nitya-consul-9hhn7 -- curl -s localhost:8500/v1/agent/checks?ns=foo | jq .
```
* Delete the pods of one of the static-client deployment
* Check that the services/checks still exist, and now the services will have metadata that they are managed by endpoints controller.
* Mark the static-server unhealthy and see that the check is updated even though it's a legacy service.

How I expect reviewers to test this PR:
Code review and if you can, the steps above.


Checklist:
